### DISCUSSION TITLE
How to let ESP32S3 PWM to keep working  during lightsleep ?

### DISCUSSION THREAD

**Discussion by stone1706** [micropython/micropython#12511] [ESP32]
Hi everyone,
 I use PWM Pin(21) to drive LED back light for my LCD .  During light sleep mode, I got a problem ,PWM was shut down ,it cannot work anymore until system got wake up. 
  How to let ESP32S3 PWM to keep working  during lightsleep ? I found there someone said PWM support LEDC feature, but I cannot find anyway to do that ...
  what is config could support it  ?
  thanks.

**Comment by scruss** [comment]
I don't think ESP32 MicroPython is able to do that currently. LEDC turns off during lightsleep with the clock modes that MicroPython uses.

**Reply by stone1706** [reply]
It's done .right now ESP32S3 PWM is keep working during lightsleep for me .
thanks .

**Comment by mattytrentini** [comment]
Is this a MicroPython issue? Is it possible to power LEDC during light sleep on the ESP32? [This Espressif forum post](https://www.esp32.com/viewtopic.php?f=12&p=49887) suggest that it's not.

If someone can supply an IDF example of powering LEDC under a sleep mode we can consider the possibility of duplicating the method in MicroPython.

**Reply by stone1706** [reply]
I got this information
  "The LED PWM module can use RTC8M_CLK as a clock source when APB_CLK is disabled. In other words,
when the system is in low-power consumption mode (see Power Management Chapter), normal peripherals will
be halted (APB_CLK is turned off), but the LED PWM can work normally via RTC8M_CLK."
code like this below
***************************************************************
      ledc_timer_config_t ledc_timer = {
        .duty_resolution = LEDC_TIMER_13_BIT, // resolution of PWM duty
        .freq_hz = 100,                      // frequency of PWM signal
        .speed_mode = LEDC_LS_MODE,           // timer mode
        .timer_num = LEDC_LS_TIMER,            // timer index
        .clk_cfg = LEDC_USE_RTC8M_CLK,              // Auto select the source clock
    };
    ledc_timer_config(&ledc_timer);

    ledc_channel_config_t ledc_channel[LEDC_TEST_CH_NUM] = {
        {
            .channel    = LEDC_LS_CH2_CHANNEL,
            .duty       = 0,
            .gpio_num   = LEDC_LS_CH2_GPIO,
            .speed_mode = LEDC_LS_MODE,
            .hpoint     = 0,
            .timer_sel  = LEDC_LS_TIMER
        },
    };
*************************************************
so... need to config pwm to use RTC8M_CLK , not APB_CLK  ???

**Reply by robert-hh** [reply]
That would be a clock change from 80M to 8M, limiting the upper frequency range of PWM to 4M, excluding it's use to create e.g. the master clock for I2S.

**Reply by stone1706** [reply]
is it possible to do that ? 

I am trying to modify machine_pwm.c file to find a way.

**Comment by stone1706** [comment]
It's done .
Config  ".clk_cfg = LEDC_USE_RTC8M_CLK" ,this is work
and now  ESP32S3 PWM is keep working during lightsleep for me .
thanks .

**Comment by arweic** [comment]
I would like to use this feature in MicroPython too. Is there a plan to implement it and if yes, when and how can I use it?
